app/uk/gov/hmrc/fhregistrationfrontend/controllers/FormPageController.scala

```diff
@@ -137,6 +138,36 @@ class FormPageController @Inject() (
         )
     }
 
+  def saveCompanyOfficers(sectionId: Option[String]): Action[AnyContent] =
+    pageAction("companyOfficers", sectionId).async { implicit request =>
+      request
+        .page[ListWithTrackedChanges[CompanyOfficer]]
+        .parseFromRequest(
+          pageWithErrors => Future successful renderForm(pageWithErrors, true),
+          page => {
+            val pageData: ListWithTrackedChanges[CompanyOfficer] = page.data.get
+            val companyOfficersPageData = pageData.values.toList
+            val usedVatNumbers: List[String] =
+              request.otherUsedVatNumbersFromCompanyOfficersPage(companyOfficersPageData, sectionId)
+            val index = sectionId.map(_.toInt - 1).getOrElse(0)
+            val vatNumberOnCompanyOfficer = CompanyOfficer.getVatNumber(companyOfficersPageData(index))
+            if (!vatNumberOnCompanyOfficer.exists(usedVatNumbers.contains)) {
+              saveSuccessfully(page)
+            } else {
+              val companyOfficersPage =
+                new InjectedPage(views).companyOfficersPage.copy(value = pageData, index = index)
+              Future successful BadRequest(
+                companyOfficersPage.renderWithFormError(
+                  CompanyOfficersForm.withError("vatRegistration", "error.vatAlreadyUsed"),
+                  request.bpr,
+                  request.journey.navigation(request.lastUpdateTimestamp, request.page)
+                )(request, request2Messages(request), appConfig)
+              )
+            }
+          }
+        )
+    }
+
```

Comment:
Continuing on from the previous comment I think this function can be made more concise and clearer if written like this:

```
def saveCompanyOfficers(sectionId: Option[String]): Action[AnyContent] =
    pageAction("companyOfficers", sectionId).async { implicit request =>
      request
        .page[ListWithTrackedChanges[CompanyOfficer]]
        .parseFromRequest(
          pageWithErrors => Future successful renderForm(pageWithErrors, true),
          page => {
            val pageData: ListWithTrackedChanges[CompanyOfficer] = page.data.get
            val companyOfficersPageData = pageData.values.toList
            val sectionIndex = sectionId.map(_.toInt - 1).getOrElse(0)
            def vatNumberIsUnique(vatNumber: String): Boolean = {
              request.isVatNumberUnique(companyOfficersPageData, vatNumber, sectionIndex)
            }
            CompanyOfficer.getVatNumber(companyOfficersPageData(sectionIndex)) match {
              case Some(vatNumber) if !vatNumberIsUnique(vatNumber) =>
                val companyOfficersPage = new InjectedPage(views)
                  .companyOfficersPage.copy(value = pageData, index = sectionIndex)

                Future successful BadRequest(
                  companyOfficersPage.renderWithFormError(
                    CompanyOfficersForm.withError("vatRegistration", "error.vatAlreadyUsed"),
                    request.bpr,
                    request.journey.navigation(request.lastUpdateTimestamp, request.page)
                  )(request, request2Messages(request), appConfig)
                )
              case _ => saveSuccessfully(page)
            }
          }
        )
    }
```

As you can see I only check the vatNumber if a vatNumber is present here.